rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+// Create a key.
+kernel_cache::key_t key = ...;
+kernel_cache::value_t value = kcache.get(key);
+
+// Assume bool(value) == true.
+gpu_value_t *gpu_value = static_cast<gpu_value_t *>(value.impl().get());
+compute::kernel_t k = gpu_value->kernel;
+// Now we can use the kernel `k`.
+```
+
+#### Memory Consumption Management
+
+oneDNN users usually want to have control over amount of memory that the library
+consumes if that amount is significant. Introducing the cache will come with
+additional memory consumption though if it is reasonably small then users may not
+want to deal with it. Given that, the are two approaches to the cache memory
+consumption.
+
+##### Option 1: Keep the Cache Hidden (recommended)
+
+This option suggests making the cache a hidden feature without providing any
+control over its capacity.
+
+There are two approaches to defining the cache capacity:
+1. Define a reasonable constant
+2. Tie the cache capacity to the primitive cache capacity
+
+The first approach is overly rigid and it is hard to define a reasonable
+capacity that would work for the majority of workloads therefore we will consider
+the second one.
+
+Pros:
+* Provides sufficient control to benefit from the cache and to maintain memory
+consumption
+* Doesn't introduce yet another *cache capacity* definition that will simplify using the cache
+
+Cons:
+* There is no guarantee that all kernels in the kernel cache are used by all the
+primitives in the primitive cache. It might happen those primitives in the
+primitive cache has unique kernels and the kernel cache may have unique kernels
+too. This means that memory consumption can be higher that it is now
+
+##### Option 2: Provide API to Control Kernel Cache Capacity
+
+This option suggests adding an API to control kernel cache capacity similar to
+the one for the primitive cache.
+
+Pros:
+* Provides more control over the kernel cache and therefore more flexibility to
+the users. Those users who heavily rely on dynamic shapes can balance between
+primitive cache capacity and the kernel cache capacity
+* User can fully control memory consumption
+
+Cons:
+* Introducing yet another *cache capacity* definition that should be specified.
+Also, we will have to explain to the users the difference between kernel cache and primitive cache capacities as well as how they interoperate with each other
+
+###### C API
+
+```c
+dnnl_status_t dnnl_get_kernel_cache_capacity(int *capacity);
+dnnl_status_t dnnl_set_kernel_cache_capacity(int capacity);
+```
+
+###### C++ API
+
+```cpp
+int get_primitive_cache_capacity();
+void set_primitive_cache_capacity(int capacity);
+```
+
+###### Environment Variable
+
+`ONEDNN_PRIMITIVE_CACHE_CAPACITY=<number>`, where `number` is the kernel cache
+capacity. The default value is 1024. When `number` is 0 the kernel cache is
+disabled.
+
+###### Build Time Control
+
+The kernel cache can be enabled or disabled at the build time using a CMake option
+`ONEDNN_ENABLE_KERNEL_CACHE`.
```

Comment:
Well, it depends on how we want to proceed with this. If we want to enable it as an experimental feature then it will be disabled otherwise it should be enabled by default. But given that I recommend the implicit kernel option it will enabled by default.